torch/_inductor/codegen/simd.py

```diff
@@ -1701,85 +1995,113 @@ def select_tiling(cls, node_schedule, numel, reduction_numel=sympy.Integer(1)):
+        # If this is a reduction, only tile reduction dims.
+        is_pointwise = reduction_numel == 1
+
+        default_tiling = cls.create_tiling([numel], [reduction_numel])
+        if config.triton.max_tiles <= 1:
+            return default_tiling
 
         seen_names: OrderedSet[str] = OrderedSet()
-        candidate_tiles: Counter[Any] = collections.Counter()
+        candidate_tiles: Counter[CandidateTiling] = collections.Counter()
         for node in EnableReduction.filter(node_schedule):
-            for tiling in cls.candidate_tilings(node):
-                if tiling.name in seen_names:
+            for candidate_tiling in cls.candidate_tilings(node, is_pointwise):
+                if candidate_tiling.name in seen_names:
                     continue
-                seen_names.add(tiling.name)
-                candidate_tiles[tiling.tiling] += tiling.score
+                elif candidate_tiling.name is not None:
+                    seen_names.add(candidate_tiling.name)
+                candidate_tiles[candidate_tiling] += candidate_tiling.score
 
-        ranked_tilings = [tiling for tiling, score in candidate_tiles.most_common()]
+        ranked_tilings: List[Dict[str, sympy.Expr]] = [
+            candidate_tiling.tiling
+            for candidate_tiling, score in candidate_tiles.most_common()
+        ]
 
-        if config.triton.max_tiles >= 3:
+        if config.triton.max_tiles >= 3 and is_pointwise:
             # Consider adding a third dimension of tiling, but only
             # when a1 is a multiple of b1; otherwise, you have a lot
             # of stragglers which is annoying to generate code for.
             #
             # NB: More than three max tiles is not enabled by default.
 
-            # Add one 3D tiling choice
-            for i in range(1, len(ranked_tilings)):
-                a0, a1 = ranked_tilings[0]
-                b0, b1 = ranked_tilings[i]
+            def convert_tiling_to_3d(
+                tiling0: Dict[str, sympy.Expr], tiling1: Dict[str, sympy.Expr]
+            ) -> Optional[Dict[str, sympy.Expr]]:
+                a0, a1 = tiling0["x"], tiling0["y"]
+                b0, b1 = tiling1["x"], tiling1["y"]
                 if V.graph.sizevars.size_hint(a1 - b1) == 0:
-                    continue
+                    return None
                 if V.graph.sizevars.size_hint(a1 - b1) < 0:
                     # swap so a0 is bigger
-                    a0, a1 = ranked_tilings[i]
-                    b0, b1 = ranked_tilings[0]
+                    (a0, a1), (b0, b1) = (b0, b1), (a0, a1)
                 assert V.graph.sizevars.size_hint(a1 - b1) > 0
-                if V.graph.sizevars.statically_known_multiple_of(a1, b1):
-                    tiling = (a0, FloorDiv(a1, b1), b1)
-                    ranked_tilings = [tiling] + ranked_tilings
+                if not V.graph.sizevars.statically_known_multiple_of(a1, b1):
+                    return None
+
+                new_tiling = {
```

Comment:
Do we have any CI tests for 3D tiling?

Reply:
Looks like it's now tested as of https://github.com/pytorch/pytorch/pull/141709/files#diff-61ac3fd3cf59657ccabd1cc334f63847e27fd3dc0bb8744cff2c93afc4563c69